tidy3d/plugins/adjoint/components/medium.py

```diff
@@ -273,40 +326,76 @@ def store_vjp(
+            for dim_index, dim_pt in enumerate("xyz"):
 
-            # compute the size of the user-supplied medium along each dimension.
-            grid = grids[eps_field_name]
-            d_sizes = grid.sizes
+                coord_dim = coords[dim_pt]
 
-            # if any of the sizes are just 0, indicating an ndim < 3, just normalize out to 1.0
-            d_sizes = (
-                np.ones(1) if len(dl) == 1 and dl[0] <= 0 else dl
-                for dl in (d_sizes.x, d_sizes.y, d_sizes.z)
-            )
+                # if it's uniform / single pixel along this dim
+                if len(np.array(coord_dim)) == 1:
+                    # discretize along this edge like a regular volume
+
+                    # compute the length of the pixel within the sim bounds
+                    r_min_coords, r_max_coords = grid.boundaries.to_list[dim_index]
+                    r_min_sim, r_max_sim = np.array(sim_bounds).T[dim_index]
+                    r_min = max(r_min_coords, r_min_sim)
+                    r_max = min(r_max_coords, r_max_sim)
+                    size = abs(r_max - r_min)
+
+                    # discretize according to PTS_PER_WVL
+                    num_cells_dim = int(size * PTS_PER_WVL_INTEGRATION / wvl_mat) + 1
+
+                    # compute the length element along the dim, handling case of sim.size=0
+                    d_len = size / num_cells_dim if size > 0 else 1.0
+                    d_sizes[dim_index] = np.array([d_len])
+
+                    # construct the interpolation coordinates along this dimension
+                    coords_interp = np.linspace(r_min + d_len / 2, r_max - d_len / 2, num_cells_dim)
+                    interp_coords[dim_pt] = coords_interp
+                    sum_axes.append(dim_pt)
+
+                # otherwise
+                else:
+                    # just evaluate at a point
+                    interp_coords[dim_pt] = coord_dim
 
             # outer product all dimensions to get a volume element mask
             d_vols = np.einsum("i, j, k -> ijk", *d_sizes)
 
-            # multiply volume element into gradient and reshape to expected vjp_shape
-            vjp_shape = tuple(len(coord) for _, coord in coords.items())
-            vjp_values = (np.squeeze(d_vols) * e_dotted.real.values).reshape(vjp_shape)
+            vjp_component = 0.0
+
+            for field_name in ("Ex", "Ey", "Ez"):
+
+                e_dotted = self.e_mult_volume(
+                    field=field_name,
+                    grad_data_fwd=grad_data_fwd,
+                    grad_data_adj=grad_data_adj,
+                    vol_coords=interp_coords,
+                    d_vol=d_vols,
+                ).sum(sum_axes)
+
+                # reshape values to the expected vjp shape to be more safe
+                vjp_shape = tuple(len(coord) for _, coord in coords.items())
+                vjp_values = e_dotted.real.values.reshape(vjp_shape)
+
+                vjp_component += vjp_values
```

Comment:
OK I am definitely confused here. So for every element of the permittivity, you sum up gradient contributions from *all three* field components? That does not make sense to me. The old handling where Ex is used for eps_xx, etc, makes more sense to me for this vjp. Maybe the problem is that the summation needs to happen at the next level, but doesn't? That is to say, when the user provides scalar epsilon, the three eps_xx, eps_yy, eps_zz gradients do need to be summed. But not here, where we still want to return the component-wise gradients, which should only depend on their corresponding field components. But maybe I'm missing something?

Reply:
No I think you might be right, actually. It was probably correct before but for some reason the example still was accurate with this new change. I need to think about it a bit more, but if the eps_ii only has direct influence over the E_i fields, I think the vjps should probably be component by component like they were before.

Reply:
I reverted this and we can continue testing with the old version.